processing/data_collection/gazette/spiders/rs_gravatai.py

```diff
@@ -0,0 +1,63 @@
+from dateparser import parse
+from datetime import datetime
+
+from scrapy import Request
+
+from gazette.items import Gazette
+from gazette.spiders.base import BaseGazetteSpider
+
+
+class RsGravataiSpider(BaseGazetteSpider):
+    TERRITORY_ID = "4309209"
+    name = "rs_gravatai"
+    allowed_domains = ["gravatai.atende.net"]
+    start_urls = ["https://gravatai.atende.net/?pg=diariooficial"]
+
+    extra_editions_options = ("Suplementar", "Retificação")
+
+    def parse(self, response):
+        """
+        @url https://gravatai.atende.net/?pg=diariooficial
+        @returns requests 68
+        """
+        last_page_number_css = "#paginacao > ul > li:nth-child(7) > button::attr(value)"
+        last_page_number = int(response.css(last_page_number_css).extract_first())
+
+        for page_number in range(1, last_page_number + 1):
+            yield Request(
+                f"https://gravatai.atende.net/?pg=diariooficial&pagina={page_number}",
+                callback=self.parse_gazette,
+            )
+
+    def parse_gazette(self, response):
+        """
+        @url https://gravatai.atende.net/?pg=diariooficial&pagina=1
+        @returns items 1
+        @scrapes date file_urls is_extra_edition territory_id power scraped_at
+        """
+
+        for element in response.css(".nova_listagem > .linha"):
+            info = element.css(".info")
+
+            is_extra_edition = (
+                info.css(".tipo::text").extract_first() in self.extra_editions_options
+            )
+
+            date = parse(
+                info.css(".data::text").extract_first(), languages=["pt"]
+            ).date()
+
+            code = element.css(".opcoes > button::attr(data-codigo)").extract_first()
+            url = (
+                "https://gravatai.atende.net/atende.php?rot=54002&aca=737&"
+                "processo=download&codigo={codigo}.pdf".format(codigo=code)
+            )
```

Comment:
Have you try returns the original URL, with the `.php`, and adjust the file name in a pipeline item? I'm not 100% sure, but I think we avoid have the `.php` and `pdf` in the `data` folder.

Reply:
I haven't tried that but I think that we can address this problem in a different way. Currently there is a pipeline [responsible for removing the query parameters](https://github.com/okfn-brasil/diario-oficial/blob/master/processing/data_collection/gazette/parser.py) from a url, which is working fine.

This pipeline is also changing the given file extension from `.pdf` to `.php` when [it overrides the media_ext variable](https://github.com/okfn-brasil/diario-oficial/blob/master/processing/data_collection/gazette/parser.py#L33) based on the path resulted from the parsing.

```python
ipdb> url
'https://gravatai.atende.net/atende.php?rot=54002&aca=737&processo=download&codigo=1053.pdf'
ipdb> urlparse(url)
ParseResult(scheme='https', netloc='gravatai.atende.net', path='/atende.php', params='', query='rot=54002&aca=737&processo=download&codigo=1053.pdf', fragment='')
```

Do you know if we have any other spiders taking advantage of this logic? I've executed the Rio de Janeiro spider but it didn't downloaded any pdf files. Because if not, I think that we may not need this if statement content any more. Since, for this specific spider, which is downloading pdf files with query parameters in its url, the media_ext variable already comes without the query parameters. What do you think?

Reply:
I have an idea. Instead of remove the pipeline item we can improve it to be more generic. I have took a glance in the scrapy source code and  noticed that the response object has the [headers](https://github.com/scrapy/scrapy/blob/a6bf895474ade06efc818582db55beb7a2545e32/scrapy/http/response/__init__.py#L21). So, we can ignore file  extensions because it does not mean to much, specially on Linux. We can use the `Content-Type` to define the file extension. If the content type is "not supported", we can use the file extension and print a warning. What do you think? :)

Furthermore, I still think you should remove that .pfd from the URL. This data is storage in the database and it can be really hard to somebody else to remember that the .pdf part is not from the original URL. Thus, we keep a better historical info. By the way, it's kind of surprise to me a server responding for a modified URL.

Reply:
Take a look on this lib as well:
```
>>> import mimetypes as mt
>>> mt.guess_all_extensions("application/pdf")
['.pdf']
```

Reply:
One more info that I found. I've already sent this to you via Telegram. But I'll put here as well for documentation. To validate my previous idea, I wrote a simple `print` statement to see the response headers:

```
{b'Date': [b'Tue, 28 Jan 2020 22:16:39 GMT'], b'Server': [b'Apache'], b'Expires': [b'Mon, 26 Jul 1997 05:00:00 GMT'], b'Cache-Control': [b'must-revalidate, post-check=0, pre-check=0'], b'Pragma': [b'public'], b'Data-Servidor': [b'1580249799000'], b'Content-Disposition': [b'attachment; filename="Edi\xe7\xe3oOrdin\xe1ria666.pdf"'], b'Strict-Transport-Security': [b'max-age=31536000; includeSubDomains; preload'], b'Content-Type': [b'application/octet-stream']} 
```

As you can see the `Content-Type` is not very helpful in this case. But it can be for other spiders. But for now, let's focus on yours. So, even if `Content-Type` does not have anything useful, because `application/octet-stream` can be anything, take a look in the `Content-Disposition`. There we can see the `filename`. Maybe is worth full try to use it to get the file type.  As a fallback we can  continue use the file extension as it is now.

Reply:
I was struggling a bit because scrapy was caching the first requests and the response was always empty, but I did manage to debug the response object content. As soon as I have some free time I'll work on it! Thanks for your support @jvanz